sbi/inference/fmpe/fmpe_base.py

```diff
@@ -0,0 +1,317 @@
+                    )
+                self.optimizer.step()
+
+            self.epoch += 1
+
+            train_loss_average = train_loss_sum / len(train_loader)  # type: ignore
+            # TODO: rename to loss once renaming is done in base class.
+            self._summary["training_log_probs"].append(-train_loss_average)
+
+            # Calculate validation performance.
+            self._neural_net.eval()
+            val_loss_sum = 0
+
+            with torch.no_grad():
+                for batch in val_loader:
+                    theta_batch, x_batch = (
+                        batch[0].to(self._device),
+                        batch[1].to(self._device),
+                    )
+                    # Aggregate the validation losses.
+                    val_losses = self._neural_net.loss(theta_batch, x_batch)
+                    val_loss_sum += val_losses.sum().item()
+
+            # Take mean over all validation samples.
+            self._val_loss = val_loss_sum / (
+                len(val_loader) * val_loader.batch_size  # type: ignore
+            )
+            # TODO: remove this once renaming to loss in base class is done.
+            self._val_log_prob = -self._val_loss
+            # Log validation log prob for every epoch.
+            # TODO: rename to loss and fix sign once renaming in base is done.
+            self._summary["validation_log_probs"].append(-self._val_loss)
+            self._summary["epoch_durations_sec"].append(time.time() - epoch_start_time)
+
+            self._maybe_show_progress(self._show_progress_bars, self.epoch)
+
+        self._report_convergence_at_end(self.epoch, stop_after_epochs, max_num_epochs)
+
+        # Update summary.
+        self._summary["epochs_trained"].append(self.epoch)
+        # TODO: rename to loss once renaming is done in base class.
+        self._summary["best_validation_log_prob"].append(self._best_val_log_prob)
+
+        # Update tensorboard and summary dict.
+        self._summarize(round_=self._round)
+
+        # Update description for progress bar.
+        if show_train_summary:
+            print(self._describe_round(self._round, self._summary))
+
+        # Avoid keeping the gradients in the resulting network, which can
+        # cause memory leakage when benchmarking.
+        self._neural_net.zero_grad(set_to_none=True)
+
+        return deepcopy(self._neural_net)
+
+    def build_posterior(
+        self,
+        density_estimator: Optional[ConditionalDensityEstimator] = None,
+        prior: Optional[Distribution] = None,
+        sample_with: str = "direct",
+        direct_sampling_parameters: Optional[Dict[str, Any]] = None,
+        **kwargs,
+    ) -> DirectPosterior:
+        """Build the posterior distribution.
+
+        Args:
+            density_estimator: Density estimator for the posterior.
+            prior: Prior distribution.
+            sample_with: Sampling method.
```

Comment:
No, for now it is not because we just doing single round using the `DirectPosterior`. Could change in a future PR when we switch to the vector field API like score matching. 

I will adapt the docstring.